docs/topics/upgrade.md

```diff
@@ -0,0 +1,77 @@
+# Upgrading Kubernetes Clusters
+
+## Prerequisites
+
+All the commands in this guide require both the Azure CLI and `aks-engine`. Follow the [quickstart guide](../tutorials/quickstart.md) before continuing.
+
+This guide assumes you already have deployed a cluster using `aks-engine`. For more details on how to do that see [deploy](../tutorials/deploy.md).
+
+## Upgrade
+
+This document provides guidance on how to upgrade the Kubernetes version for an existing `aks-engine` cluster and recommendations for adopting `aks-engine upgrade` as a production quality tool.
+
+<a name="pre-requirements"></a>
+
+### Know before you go
+
+In order to ensure that your `aks-engine upgrade` operation runs smoothly, there are a few things you should be aware of before getting started.
+
+1) You will need access to the original `_output/` directory files that were generated during cluster deployment in order to run `upgrade`.  `aks-engine` will use the `--deployment-dir` argument provided in the upgrade command to introspect the Azure Resource Manager (ARM) template and parameters to determine the cluster's current Kubernetes version and its current state.
```

Comment:
Strictly speaking, aks-engine is introspecting the apimodel.json, right? So we should probably clarify that.

"... to introspect the `apimodel.json` cluster configuration, in order to determine the cluster's current Kubernetes version, as well as all other cluster configuration data as defined by aks-engine during the last time that aks-engine was used to create, scale, or upgrade the cluster."

Let's leave out the determination of state, as that's not 100% correct.

Reply:
no, for upgrade aks-engine introspects the `azuredeploy.json` and the parameters file

Reply:
will change the state part of the sentence

Reply:
Hmm, that's surprising. Does it reconcile any differences that it detects between apimodel.json and azuredeploy.json?

Reply:
correction ** it does use the apimodel to load the container service and saves the file at the end with updated k8s version but it does use the azuredeploy json and parameters for fetching the name suffix

Reply:
Just to be clear, upgrade is doing some local data gathering so it knows "what the resource profile in the RG looks like" <-- something like that, right?

My question is: *should* it be relying upon the template output for that data gathering pre-step? It seems to be an anti-pattern. What's preventing upgrade from referring only to the apimodel.json (i.e., refererring only to a data representation that can be coupled with an in-code, strongly typed data model)?

Reply:
https://github.com/Azure/aks-engine/issues/361